### ISSUE TITLE
Callbacks or trait impl's for allowing user to track events

### ISSUE THREAD

**Issue by DevinR528** [matrix-org/matrix-rust-sdk#3]
For clients that want `matrix-sdk` to handle most of the event and room mutation, only wanting to be notified when something changes certain events a trait could be implemented that tracks all of the events and calls methods from the users struct if it impl's this trait.
```rust
/// And in `AsyncClient::sync` there could be a switch case that calls the corresponding method on
/// the `Box<dyn EventEmitter>
pub trait EventEmitter {
    /// Event that alters the name or alias of this room.
    fn on_room_name(&mut self, _: &Room, _: &RoomEvent) {}
    /// Any event that alters the state of the room's members.
    fn on_room_member(&mut self, _: &Room, _: &RoomEvent) {}
}

pub struct MyApp {
    member: Vec<RoomMember>,
    ...
}

impl MyApp {
    fn update_ui(&mut self, members: &HashMap<UserId, RoomMember>) {
        self.members.push(members.get())
        ...
    }
}

impl EventEmitter for MyApp {
    fn on_room_member(&mut self, room: &Room, ev: &RoomEvent) {
        // allowing something like this
        match ev.membership_change() {
            MembershipState::Joined => self.update_ui(&room.members))
        }
    }
    // Methods for all relevant events .
}

async fn main() {
    let client = AsyncClient::with_emitter(Box::new(MyApp));
    loop {
        draw_ui();
        // calling sync will call the proper methods of MyApp
        client.sync();
    }
}
```

**Comment by poljar** [comment]
I thought about this over the weekend and I think the proposal makes sense, otherwise, depending on the callback granularity we want, we might end up with way to many `add_callback_X()` methods.

While requiring to implement a trait for the callbacks will make the simple responder bot more complex it'll probably make anything more sophisticated easier.

Couple things to note:
* The callbacks need to return futures, that is you're callback needs to be able to call `client.room_send().await` directly, otherwise making a simple responder bot will be unnecessarily hard.
* We need to support multi-threaded executors so the trait will probably depend on the Send + Sync trait bounds.
* I don't think that borrows for the events in the callback will work because of the async requirement, happy to be proven wrong though.
* We probably want to be able to add this trait object at any point in time, not just at client creation.

I'm still a bit undecided how many callback types we want we probably want something better than just one for all `RoomEvent` variants but considering how [many](https://docs.rs/ruma-events/0.18.0/ruma_events/collections/all/enum.RoomEvent.html) event types are in the `RoomEvent` enum we probably don't want one per event.

Another thing, a bit unrelated to this but since you put it in your example:
```rust
room.members.new_member()
```
I'm not sure how this should work, isn't this the same as [this](https://docs.rs/ruma-events/0.18.0/ruma_events/room/member/struct.MemberEvent.html#method.membership_change) method and checking for the `Joined` enum variant?

And lastly, please make sure to create a separate PR for this, since it'll change a lot of things around to ease the review a bit.

**Comment by DevinR528** [comment]
> I'm not sure how this should work, isn't this the same as this method and checking for the Joined enum variant?

Yea that is basically the same. I wonder about making the signature 
```rust
#[async_traits]
pub trait EventEmitter: Send + Sync {
    /// Event that alters the name or alias of this room.
    async fn on_room_name(&mut self, did_update: bool, _: Arc<Room>, _: Arc<RoomEvent>)
        -> Result<()>
    {}
    ...
}
```
Could we add [async_traits](https://github.com/dtolnay/async-trait) or should I just write out the return using a `Future`? Oh and your definitely right about `&Room` and `&RoomEvent` not working.  

> I'm still a bit undecided how many callback types we want we probably want something better than just one for all RoomEvent variants but considering how many event types are in the RoomEvent enum we probably don't want one per event.

I added one for `PresenceEvent` yesterday and realized the same thing (how many seperate events their are). We could start with the events that makes sense for our use cases and add as needed.

In my head I see the `AsyncClient` as having two jobs, first set up the initial state of the room like our `get_room_name` methods, second keeping track of events that are going to happen, this is the callbacks/EventEmitter trait. So after logging in and syncing you can ask the client for the current state of the rooms then rely on the trait methods to keep everything updated. Is this sort of how you see it?

The initial state the AsyncClient should collect and have methods for
- room name
- room members, the `HashMap<UserId, RoomMember>`
- push ruleset
- ignored users
- current room id

These are kept up to date by the trait
- room name
- user appearance stuff
- user membership 
- messages
- `IncomingEphemeral` like typing events
- what ever else is needed

**Comment by poljar** [comment]
> Could we add [async_traits](https://github.com/dtolnay/async-trait) or should I just write out the return using a `Future`? Oh and your definitely right about `&Room` and `&RoomEvent` not working.

We already depend on async_traits [here](https://github.com/matrix-org/matrix-rust-sdk/blob/master/Cargo.toml#L15), though the errors async trait spits out are a bit iffy, hopefully that'll get better.

> I added one for `PresenceEvent` yesterday and realized the same thing (how many seperate events their are). We could start with the events that makes sense for our use cases and add as needed.

Agree. Lets stick to a couple first, presence, room_event, to-device event, ephemeral event. And as needed we might add granularity to the room_event ones.


> In my head I see the `AsyncClient` as having two jobs, first set up the initial state of the room like our `get_room_name` methods, second keeping track of events that are going to happen, this is the callbacks/EventEmitter trait. So after logging in and syncing you can ask the client for the current state of the rooms then rely on the trait methods to keep everything updated. Is this sort of how you see it?

The client should keep the state in sync the whole time the `sync_forever()` method is running, the callbacks are here to notify the owner of the client that something changed, that is that events were received that change the state (you probably meant this as well but I hope I stated it a bit more explicitly). The rest is correct, yes.

> Could I get #2 to a merge-able state so I can start from there with this PR?

If #2 is ready for review just yell, and I'll review it soon™. And yeah it's fine if this PR depends on #2, they probably build upon each other since you added presence callbacks there.

**Comment by DevinR528** [comment]
I think something like this would be useful for how I'm thinking of implementing the event loop
```rust
/// Synchronise the client's state with the latest state on the server using the `EventEmitter`
/// trait.
///
/// # Arguments
///
/// * `sync_settings` - Settings for the sync call.
/// * `event_emitter` - implementors of the `EventEmitter` trait to handle state callbacks.
// #[instrument]
pub async fn sync_with(
    &mut self,
    sync_settings: SyncSettings,
    event_emitter: Arc<Mutex<dyn EventEmitter>>,
) -> Result<()> {
    self.event_emitter = Some(event_emitter);

    let result = self.sync(sync_settings).await.map(|_res| ());
    
    self.event_emitter.take();
    result
}
```
Having this option and the ability to instantiate the `AsyncClient` with one or as you said a builder like method to add the `EventEmitter`?

**Comment by poljar** [comment]
Ah no, that's not what I meant. A couple of people that use nio have found it to be quite useful to register event callbacks (or in our case the event emitter) only after some time the client is already running, mostly after the first/initial sync.

This is so their bots can ignore everything that isn't a live message. So we would need a `add_event_emitter()` method and the `sync()` method checks if a event emitter exists, if not do nothing, if yes call the callbacks of it.

Isn't that method running the sync only once as well? Don't you want to run the `sync_forever()` method instead? Correctly using encryption means that we're gonna need to do some additional requests between syncs, that's what the `sync_forever()` method does for you.

Why do you need to add an event emitter, sync once, remove the event emitter? Sure we'll probably want to have a way to remove the event emitter but not sure why we would want to do this after every sync?

Heh your loops look a bit similar to mine in [Weechat](https://github.com/poljar/weechat-matrix-rs/blob/master/src/server.rs#L558), though I won't be using an event emitter since I have to deal with Weechat having a strictly single threaded API.

**Comment by DevinR528** [comment]
Yea now that you say it, it's dumb to add remove add remove, and basically rewriting `sync_forever` doesn't make much sense. So I'll stick with an `add_event_emitter` method and when I actually start this PR `sync` will work as it does just using the EventEmitter trait.

**Comment by poljar** [comment]
One thing I forgot to mention. The base client used to be non-async, that's why all the callback handling ended up in the `AsyncClient`. We might want to move our callbacks and handling of them into the base client.

If it's not clear what the distinction between them is, the base client is a pure state machine that receives responses and tells you what requests you need to do, in the case of encryption support at least.

The AsyncClient should live on top of it and do requests via some transport, HTTP in our case.

So the base client could handle all the callbacks, that way using differing request libraries if people wish to do so should be even easier.

**Comment by poljar** [comment]
This has been implemented in #5.
